src/OrbitBase/include/OrbitBase/Future.h

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_FUTURE_H_
+#define ORBIT_BASE_FUTURE_H_
+
+#include <absl/synchronization/mutex.h>
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <variant>
+
+#include "OrbitBase/Result.h"
+#include "OrbitBase/SharedState.h"
+
+namespace orbit_base_internal {
+
+// For befriending Promise we need a forward declaration.
+template <typename T>
+class PromiseBase;
+
+// FutureBase<T> is an internal base class for Future<T>. Check out Future<T> below for more
+// information.
+template <typename T>
+class FutureBase {
+ public:
+  FutureBase(const FutureBase&) = default;
+  FutureBase& operator=(const FutureBase&) = default;
+  FutureBase(FutureBase&&) = default;
+  FutureBase& operator=(FutureBase&&) = default;
+  ~FutureBase() noexcept = default;
+
+  [[nodiscard]] bool IsValid() const { return shared_state_.use_count() > 0; }
+
+ protected:
+  explicit FutureBase(std::shared_ptr<SharedState<T>> shared_state)
+      : shared_state_{std::move(shared_state)} {}
+
+  std::shared_ptr<SharedState<T>> shared_state_;
+};
+
+}  // namespace orbit_base_internal
+
+namespace orbit_base {
+
+/* A future type, similar to std::future but prepared to integrate with
+   Orbit-specific code like MainThreadExecutor and ThreadPool.
+
+   Future is a result type for an asynchronous task, where the result
+   will not be available right away, but to a later point in time.
+
+   A valid Future<T> is created from a Promise<T>. The latter lives in the
+   asynchronous task and its purpose is to notify the Future<T> when the result
+   is available.
+
+   Use Future<T>::IsValid() to check if a future is connected to a promise or holds
+   a result value.
+
+   Call Future<T>::IsFinished() to figure out if the result is already available.
+   If yes, you can call Future<T>::Get() to retrieve the T.
```

Comment:
I would have expected that calling Get() would block until the result is available like std::future<T>::get. Does this mean that the user always has to implement a loop to check IsFinished() before getting the result?

Reply:
Good point. I will change that to a blocking behaviour!

Reply:
Done.

Reply:
Awesome, thanks!